provider/aws_sd.go

```diff
@@ -0,0 +1,626 @@
+	config := aws.NewConfig()
+
+	config = config.WithHTTPClient(
+		instrumented_http.NewClient(config.HTTPClient, &instrumented_http.Callbacks{
+			PathProcessor: func(path string) string {
+				parts := strings.Split(path, "/")
+				return parts[len(parts)-1]
+			},
+		}),
+	)
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            *config,
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, err
+	}
+	sess.Handlers.Build.PushBack(request.MakeAddToUserAgentHandler("ExternalDNS", externaldns.Version))
+
+	provider := &AWSSDProvider{
+		client:              sd.New(sess),
+		namespaceFilter:     domainFilter,
+		namespaceTypeFilter: namespaceTypeFilter.toAwsAPIRequestFilter(),
+		dryRun:              dryRun,
+		ownerID:             ownerID,
+	}
+
+	return provider, nil
+}
+
+// Records returns list of all endpoints.
+func (p *AWSSDProvider) Records() (endpoints []*endpoint.Endpoint, err error) {
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ns := range namespaces {
+		services, err := p.ListServicesByNamespaceID(ns.Id)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, srv := range services {
+			instances, err := p.ListInstancesByServiceID(srv.Id)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(instances) > 0 {
+				// DNS name of the record is a concatenation of service and namespace
+				dnsName := *srv.Name + "." + *ns.Name
+				ep, err := p.instancesToEndpoint(dnsName, srv, instances)
+				if err != nil {
+					return nil, err
+				}
+				endpoints = append(endpoints, ep)
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (p *AWSSDProvider) instancesToEndpoint(recordName string, srv *sd.Service, instances []*sd.InstanceSummary) (*endpoint.Endpoint, error) {
+	newEndpoint := &endpoint.Endpoint{
+		DNSName:   recordName,
+		RecordTTL: endpoint.TTL(aws.Int64Value(srv.DnsConfig.DnsRecords[0].TTL)),
+		Targets:   make(endpoint.Targets, 0, len(instances)),
+	}
+
+	for _, inst := range instances {
+		// CNAME
+		if inst.Attributes["AWS_INSTANCE_CNAME"] != nil && aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeCname {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_CNAME"]))
+
+			// ALIAS
+		} else if inst.Attributes["AWS_ALIAS_DNS_NAME"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_ALIAS_DNS_NAME"]))
+
+			// IP-based target
+		} else if inst.Attributes["AWS_INSTANCE_IPV4"] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeA
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes["AWS_INSTANCE_IPV4"]))
+		} else {
+			return nil, fmt.Errorf("FATAL ERROR: Instance not valid \"%v\"", inst)
```

Comment:
why is this error necessary, can we just skip it ?

Reply:
Sure 👍 Logging as a warning will be enough.